File: sdk/dataflux_func_sdk.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import time
import uuid
import json
import hmac
from hashlib import sha1, md5
import types

try:
    import httplib
    from urllib import urlencode, quote
    from urlparse import urlparse
except ImportError:
    import http.client as httplib
    from urllib.parse import urlparse, urlencode, quote

UPLOAD_DISABLED = False
try:
    import requests
except ImportError:
    UPLOAD_DISABLED = True
    print('Cannot import `requests` module, the DataFluxFunc.upload() method is disabled.')

COLOR_MAP = {
    'grey'   : '\033[0;30m',
    'red'    : '\033[0;31m',
    'green'  : '\033[0;32m',
    'yellow' : '\033[0;33m',
    'blue'   : '\033[0;34m',
    'magenta': '\033[0;35m',
    'cyan'   : '\033[0;36m',
}

# -----------------------------------------------
# Python2 ~ Python3 Compatibility Code From `six`
# -----------------------------------------------
#
# Copyright (c) 2010-2019 Benjamin Peterson
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3
PY34 = sys.version_info[0:2] >= (3, 4)

if PY3:
    string_types = str,
    integer_types = int,
    class_types = type,
    text_type = str
    binary_type = bytes
else:
    string_types = basestring,
    integer_types = (int, long)
    class_types = (type, types.ClassType)
    text_type = unicode
    binary_type = str

def ensure_binary(s, encoding='utf-8', errors='strict'):
    if isinstance(s, text_type):
        return s.encode(encoding, errors)
    elif isinstance(s, binary_type):
        return s
    else:
        raise TypeError("not expecting type '%s'" % type(s))

def ensure_str(s, encoding='utf-8', errors='strict'):
    if not isinstance(s, (text_type, binary_type)):
        raise TypeError("not expecting type '%s'" % type(s))
    if PY2 and isinstance(s, text_type):
        s = s.encode(encoding, errors)
    elif PY3 and isinstance(s, binary_type):
        s = s.decode(encoding, errors)
    return s

def colored(s, color=None):
    if not color:
        color = 'yellow'

    color = COLOR_MAP[color]

    return color + '{}\033[0m'.format(s)

def json_dumps(obj):
    return json.dumps(obj, sort_keys=True, ensure_ascii=False, separators=(',', ':'))

class DataFluxFunc(object):
    def __init__(self, ak_id=None, ak_secret=None, host=None, port=None, timeout=3, use_https=False, debug=False):
        self.debug = debug or False

        self.ak_id     = ak_id
        self.ak_secret = ak_secret

        self.host = host
        self.port = port
        if self.host and ':' in self.host:
            host_port = self.host.split(':')
            self.host = host_port[0]
            self.port = int(host_port[1])

        self.timeout   = timeout
        self.use_https = use_https

        if not self.port:
            if self.use_https:
                self.port = 443
            else:
                self.port = 80

    def get_body_md5(self, body=None):
        body = json_dumps(body or {})

        if self.debug:
            print('{} {}'.format(
                colored('[Body to MD5]'),
                body))

        h = md5()
        h.update(ensure_binary(body))

        return h.hexdigest()

    def get_sign(self, method, path, timestamp, nonce, body=None):
        method = method.upper()

        if timestamp is None:
            timestamp = str(int(time.time()))

        if nonce is None:
            nonce = str(uuid.uuid4().hex)

        body_md5 = self.get_body_md5(body)
        string_to_sign = '&'.join([ method, path, timestamp, nonce, body_md5 ])

        if self.debug:
            print('{} {}'.format(
                colored('[String to Sign]'),
                string_to_sign))

        h = hmac.new(ensure_binary(self.ak_secret), ensure_binary(string_to_sign), sha1)
        sign = h.hexdigest()

        if self.debug:
            print('{} {}'.format(
                colored('[Signature]'),
                sign))

        return sign

    def verify_sign(self, sign, method, path, timestamp, nonce, body=None):
        expected_sign = self.get_sign(method, path, timestamp, nonce, body=body)

        return (sign == expected_sign)

    def get_auth_header(self, method, path, body=None):
        timestamp = str(int(time.time()))
        nonce     = uuid.uuid4().hex

        sign = self.get_sign(method, path, timestamp, nonce, body=body)

        auth_headers = {
            'X-Dff-Ak-Id'       : self.ak_id,
            'X-Dff-Ak-Timestamp': timestamp,
            'X-Dff-Ak-Nonce'    : nonce,
            'X-Dff-Ak-Sign'     : sign,
        }
        return auth_headers

    def verify_auth_header(self, headers, method, path, body=None):
        _headers = dict([ (k.lower(), v) for k, v in headers.items() ])

        sign      = headers.get('x-dff-ak-sign')      or ''
        timestamp = headers.get('x-dff-ak-timestamp') or ''
        nonce     = headers.get('x-dff-ak-nonce')     or ''

        return self.verify_sign(sign, method, path, timestamp, nonce, body=body)

    def run(self, method, path, query=None, body=None, headers=None, trace_id=None):
        # Prepare method / query / body
        method = method.upper()

        if query:
            path = path + '?' + urlencode(query)

        dumped_body = json_dumps(body)

        if self.debug == True:
            print('=' * 50)
            print('{} {} {}'.format(colored('[Request]'), colored(method, 'cyan'), colored(path, 'cyan')))
            if body:
                print('{} {}'.format(colored('[Body]'), dumped_body))

        # Prepare headers with auth info
        headers = headers or {}
        headers['Content-Type'] = 'application/json'
        if trace_id:
            headers['X-Trace-Id'] = trace_id

        if self.ak_id and self.ak_secret:
            auth_headers = self.get_auth_header(method, path, body=body)
            headers.update(auth_headers)

        # Do HTTP / HTTPS
        conn = None
        if self.use_https:
            conn = httplib.HTTPSConnection(self.host, port=self.port, timeout=self.timeout)
        else:
            conn = httplib.HTTPConnection(self.host, port=self.port, timeout=self.timeout)

        if body:
            conn.request(method, path, body=ensure_binary(dumped_body), headers=headers)
        else:
            conn.request(method, path, headers=headers)

        # Get response
        resp = conn.getresponse()

        resp_status_code = resp.status
        resp_raw_data    = resp.read()

        resp_content_type = resp.getheader('Content-Type')
        if isinstance(resp_content_type, string_types):
            resp_content_type = resp_content_type.split(';')[0].strip()

        if resp_content_type == 'application/json':
            resp_data = json.loads(resp_raw_data)
        else:
            resp_data = resp_raw_data

        if self.debug == True:
            _color = 'green'
            if resp_status_code >= 400:
                _color = 'red'

            print(colored('{} {}'.format('[Response]', resp_status_code), _color))
            print(colored('{} {}'.format('[Body]', ensure_str(resp_raw_data)), _color))

        return resp_status_code, resp_data

    def upload(self, path, file_buffer, filename=None, query=None, fields=None, headers=None, trace_id=None):
        if UPLOAD_DISABLED:
            raise Exception('`DataFluxFunc.upload()` method need `requests` module.')

        # Prepare method/query/fields/file
        method = 'POST'

        if not filename:
            filename = 'uploadfile'

        filename = quote(filename)

        if query:
            path = path + '?' + urlencode(query)

        if fields and not isinstance(fields, dict):
            raise Exception('`fields` should be a plain JSON')

        dumped_fields = json_dumps(fields)

        files = {'files': (filename, file_buffer)}

        if self.debug == True:
            print('=' * 50)
            print('{} {} {}'.format(colored('[Request]'), colored(method, 'cyan'), colored(ensure_str(path), 'cyan')))
            if fields:
                print('{} {}'.format(colored('[Fields]'), dumped_fields))
            if files:
                print('{} {}'.format(colored('[File]'), filename))

        # Preapre HTTP
        url = 'http://{}:{}{}'.format(self.host, self.port, path)
        req = requests.Request('POST', url, data=fields, files=files)
        prepared_req = req.prepare()

        # Prepare headers with auth info
        headers = headers or {}
        if trace_id:
            headers['X-Trace-Id'] = trace_id

        if self.ak_id and self.ak_secret:
            auth_headers = self.get_auth_header(method, path, body=fields)
            headers.update(auth_headers)

        prepared_req.headers.update(headers)

        with requests.Session() as s:
            resp = s.send(prepared_req)

        # Get response
        resp_status_code = resp.status_code
        resp_raw_data    = resp.text

        resp_content_type = resp.headers.get('Content-Type') or ''
        if isinstance(resp_content_type, string_types):
            resp_content_type = resp_content_type.split(';')[0].strip()

        if resp_content_type == 'application/json':
            resp_data = json.loads(resp_raw_data)
        else:
            resp_data = resp_raw_data

        if self.debug == True:
            _color = 'green'
            if resp_status_code >= 400:
                _color = 'red'

            print(colored('{} {}'.format('[Response]', resp_status_code), _color))
            print(colored('{} {}'.format('[Body]', ensure_str(resp_raw_data)), _color))

        return resp_status_code, resp_data

    def get(self, path, query=None, headers=None, trace_id=None):
        return self.run('GET', path, query, None, headers, trace_id=trace_id)

    def post(self, path, query=None, body=None, headers=None, trace_id=None):
        return self.run('POST', path, query, body, headers, trace_id=trace_id)

    def put(self, path, query=None, body=None, headers=None, trace_id=None):
        return self.run('PUT', path, query, body, headers, trace_id=trace_id)

    def delete(self, path, query=None, headers=None, trace_id=None):
        return self.run('DELETE', path, query, None, headers, trace_id=trace_id)

if __name__ == '__main__':
    from dataflux_func_sdk import DataFluxFunc

    # Create DataFlux Func Handler
    host = 'localhost:8088'
    if len(sys.argv) >= 2:
        host = sys.argv[1]

    dff = DataFluxFunc(ak_id='ak-xxxxx', ak_secret='xxxxxxxxxx', host=host)

    # Debug ON
    dff.debug = True

    # Send GET Request
    try:
        status_code, resp = dff.get('/api/v1/do/ping')

    except Exception as e:
        print(colored(e, 'red'))
        raise

    # Send POST Request
    try:
        body = {
            'echo': {
                'int'    : 1,
                'str'    : 'Hello World',
                'unicode': u'你好，世界！',
                'none'   : None,
                'boolean': True,
            }
        }
        status_code, resp = dff.post('/api/v1/do/echo', body=body)

    except Exception as e:
        print(colored(e, 'red'))
        raise

    # Send UPLOAD Request
    if not UPLOAD_DISABLED:
        try:
            filename = __file__
            with open(filename, 'rb') as _f:
                file_buffer = _f.read()

            fields = {
                'folder': 'test'
            }

            status_code, resp = dff.upload('/api/v1/resources/do/upload',
                file_buffer=file_buffer,
                filename=filename,
                fields=fields)

        except Exception as e:
            print(colored(e, 'red'))
            raise

File: sdk/dataflux_func_sdk.js
'use strict';

var fs          = require('fs');
var crypto      = require('crypto');
var http        = require('http');
var https       = require('https');
var querystring = require('querystring');

var UPLOAD_DISABLED = false;
try {
 var FormData = require('form-data');
} catch(err) {
  UPLOAD_DISABLED = true;
  console.log('Cannot require `form-data` package, the DataFluxFunc.upload() method is disabled.')
}

var COLOR_MAP = {
  'grey'   : '\x1b[0;30m',
  'red'    : '\x1b[0;31m',
  'green'  : '\x1b[0;32m',
  'yellow' : '\x1b[0;33m',
  'blue'   : '\x1b[0;34m',
  'magenta': '\x1b[0;35m',
  'cyan'   : '\x1b[0;36m',
}

function strf() {
  var args = Array.prototype.slice.call(arguments);
  if (0 === args.length) {
    return '';
  }

  var pattern = args.shift();
  try {
    pattern = pattern.toString();
  } catch (err) {
    pattern = '';
  }

  return pattern.replace(/\{(\d+)\}/g, function replaceFunc(m, i) {
    return args[i] + '';
  });
};

function colored(s, color) {
  if (!color) {
    color = 'yellow';
  }

  color = COLOR_MAP[color];

  return color + s + '\x1b[0m';
};

function jsonDumps(obj) {
  if (typeof obj == 'string') {
    return '"' + obj + '"';
  }
  else if (typeof obj == 'number') {
    return obj;
  }
  else if (typeof obj == 'boolean') {
    return obj;
  }
  else if (typeof obj == 'function') {
    return '"<FUNCTION>"';
  }
  else if (typeof obj == 'object') {
    if (Array.isArray(obj)) {
      var parts = [];
      for (var i = 0; i < obj.length; i++) {
        var v = jsonDumps(obj[i]);
        parts.push(v);
      }

      return '[' + parts.join(',') + ']';
    } else if (obj === null) {
      return 'null';
    } else {
      var keyList = [];
      for (var k in obj) {
        keyList.push(k);
      }
      keyList.sort();

      var parts = [];
      for (var i = 0; i < keyList.length; i++) {
        var k = keyList[i];
        var v = jsonDumps(obj[k]);

        parts.push('"' + k + '":' + v);
      }
      return '{' + parts.join(',') + '}';
    }
  }
  else {
    return '"<UNKNOW>"';
  }
};

var DataFluxFunc = function(options) {
  this.debug = options.debug || false;

  this.akId     = options.akId;
  this.akSecret = options.akSecret;

  this.host = options.host;
  this.port = options.port;

  if (this.host && this.host.indexOf(':') > -1) {
    var host_port = this.host.split(':');
    this.host = host_port[0];
    this.port = parseInt(host_port[1]);
  }

  this.timeout  = options.timeout  || 3;
  this.useHTTPS = options.useHTTPS || false;

  if (!this.port) {
    if (this.useHTTPS) {
      this.port = 443;
    } else {
      this.port = 80;
    }
  }
};

DataFluxFunc.prototype.getBodyMD5 = function(body) {
  body = jsonDumps(body || {});

  if (this.debug) {
    console.log(strf('{0} {1}',
      colored('[Body to MD5]'),
      body));
  }

  var c = crypto.createHash('md5');
  c.update(body);

  var hash = c.digest('hex');
  return hash;
};

DataFluxFunc.prototype.getSign = function(method, path, timestamp, nonce, body) {
  method = method.toUpperCase();

  if (!timestamp) {
    timestamp = parseInt(Date.now() / 1000).toString();
  }

  if (!nonce) {
    nonce = Math.random().toString();
  }

  var bodyMD5 = this.getBodyMD5(body);
  var stringToSign = [ method, path, timestamp, nonce, bodyMD5 ].join('&');

  if (this.debug) {
    console.log(strf('{0} {1}',
      colored('[String to Sign]'),
      stringToSign));
  }

  var c = crypto.createHmac('sha1', this.akSecret);
  c.update(stringToSign);
  var sign = c.digest('hex');

  if (this.debug) {
    console.log(strf('{0} {1}',
      colored('[Signature]'),
      sign));
  }

  return sign;
};

DataFluxFunc.prototype.verifySign = function(sign, method, path, timestamp, nonce, body) {
  var expectedSign = this.getSign(method, path, timestamp, nonce, body);

  return (sign === expectedSign);
};

DataFluxFunc.prototype.getAuthHeader = function(method, path, body) {
  var timestamp = parseInt(Date.now() / 1000).toString();
  var nonce     = Math.random().toString();

  var sign = this.getSign(method, path, timestamp, nonce, body);

  var authHeader = {
    'X-Dff-Ak-Id'       : this.akId,
    'X-Dff-Ak-Timestamp': timestamp,
    'X-Dff-Ak-Nonce'    : nonce,
    'X-Dff-Ak-Sign'     : sign,
  };
  return authHeader;
};

DataFluxFunc.prototype.verifyAuthHeader = function(headers, method, path, body) {
  var _headers = {};
  for (let k in headers) {
    _headers[k.toLowerCase()] = headers[k];
  }

  var sign      = _headers['x-dff-ak-sign']      || '';
  var timestamp = _headers['x-dff-ak-timestamp'] || '';
  var nonce     = _headers['x-dff-ak-nonce']     || '';

  return this.verifySign(sign, method, path, timestamp, nonce, body);
};

DataFluxFunc.prototype.run = function(options, callback) {
  var method  = options.method || 'GET';
  var path    = options.path;
  var query   = options.query;
  var body    = options.body;
  var headers = options.headers;
  var traceId = options.traceId;

  // Prepare method / query / body
  method = method.toUpperCase();

  if (query) {
    path = path + '?' + querystring.stringify(query);
  }

  var dumpedBody = jsonDumps(body);

  if (this.debug) {
    console.log('='.repeat(50));
    console.log(strf('{0} {1} {2}', colored('[Request]'), colored(method, 'cyan'), colored(path, 'cyan')));
    if (body) {
      console.log(strf('{0} {1}', colored('[Body]'), dumpedBody));
    }
  }

  // Prepare headers with auth info
  headers = headers || {};
  headers['Content-Type'] = 'application/json';
  if (traceId) {
    headers['X-Trace-Id'] = traceId;
  }

  if (this.akId && this.akSecret) {
    var authHeader = this.getAuthHeader(method, path, body);
    Object.assign(headers, authHeader);
  }

  // Do HTTP / HTTPS
  var requestOptions = {
    host   : this.host,
    port   : this.port,
    path   : path,
    method : method,
    headers: headers,
    timeout: this.timeout,
  };

  var httpLib = this.useHTTPS ? https : http;

  var self = this;

  var respStatusCode = 0;
  var respRawData    = '';
  var respData       = '';
  var req = httpLib.request(requestOptions, function(res) {
    respStatusCode = res.statusCode;

    res.on('data', function(chunk) {
      respRawData += chunk;
    });

    res.on('end', function() {
      var respContentType = res.headers['content-type'].split(';')[0].trim();
      if (respContentType === 'application/json') {
        respData = JSON.parse(respRawData);
      }

      if (self.debug) {
        var _color = respStatusCode >= 400 ? 'red' : 'green';

        console.log(colored(strf('{0} {1}', '[Response]', respStatusCode), _color));
        console.log(colored(strf('{0} {1}', '[Body]', respRawData.toString()), _color));
      }

      if ('function' === typeof callback) {
        return callback(null, respData, respStatusCode);
      }
    });
  });

  req.on('error', function(err) {
    if ('function' === typeof callback) {
      return callback(err);
    }
  });

  if (body) {
    req.write(dumpedBody);
  }

  req.end();
};

DataFluxFunc.prototype.upload = function(options, callback) {
  if (UPLOAD_DISABLED) {
    throw Error('`DataFluxFunc.upload()` method needs `form-data` package.')
  }

  var path        = options.path;
  var fileBuffer  = options.fileBuffer;
  var filename    = options.filename;
  var query       = options.query;
  var fields      = options.fields;
  var headers     = options.headers;
  var traceId     = options.traceId;

  // Prepare method / query / fields / file
  var method = 'POST';

  if (query) {
    path = path + '?' + querystring.stringify(query);
  }

  if (fields && 'object' !== typeof fields) {
    throw Error('`fields` should be a plain JSON')
  }

  // FormData
  var formData = new FormData();

  if (fields) {
    for (var k in fields) if (fields.hasOwnProperty(k)) {
      formData.append(k, fields[k]);
    }
  }

  filename = querystring.escape(filename || 'uploadfile');
  formData.append('files', fileBuffer, { filename: filename });

  if (this.debug) {
    console.log('='.repeat(50));
    console.log(strf('{0} {1} {2}', colored('[Request]'), colored(method, 'cyan'), colored(path, 'cyan')));
    if (fields) {
      console.log(strf('{0} {1}', colored('[Fields]'), fields));
    }
    if (fileBuffer) {
      console.log(strf('{0} {1}', colored('[File]'), filename));
    }
  }

  // Prepare headers with auth info
  headers = headers || {};
  headers['Content-Type'] = 'application/json';
  if (traceId) {
    headers[this.headerFields.traceId] = traceId;
  }

  if (this.akId && this.akSecret) {
    var authHeader = this.getAuthHeader(method, path, fields);
    Object.assign(headers, authHeader);
  }

  var formHeader = formData.getHeaders();
  Object.assign(headers, formHeader);

  // Do HTTP/HTTPS
  var requestOptions = {
    host   : this.host,
    port   : this.port,
    path   : path,
    method : method,
    headers: headers,
    timeout: this.timeout,
  };

  var httpLib = this.useHTTPS ? https : http;

  var self = this;

  var respStatusCode = 0;
  var respRawData    = '';
  var respData       = '';
  var req = httpLib.request(requestOptions, function(res) {
    respStatusCode = res.statusCode;

    res.on('data', function(chunk) {
      respRawData += chunk;
    });

    res.on('end', function() {
      var respContentType = res.headers['content-type'].split(';')[0].trim();
      if (respContentType === 'application/json') {
        respData = JSON.parse(respRawData);
      }

      if (self.debug) {
        var _color = respStatusCode >= 400 ? 'red' : 'green';

        console.log(colored(strf('{0} {1}', '[Response]', respStatusCode), _color));
        console.log(colored(strf('{0} {1}', '[Body]', respRawData.toString()), _color));
      }

      if ('function' === typeof callback) {
        return callback(null, respData, respStatusCode);
      }
    });
  });

  req.on('error', function(err) {
    if ('function' === typeof callback) {
      return callback(err);
    }
  });

  if (formData.getLengthSync() > 0) {
    formData.pipe(req);
  }

  req.end();
};

DataFluxFunc.prototype.get = function(options, callback) {
  options = options || {};
  options.method = 'GET';
  return this.run(options, callback);
};

DataFluxFunc.prototype.post = function(options, callback) {
  options = options || {};
  options.method = 'POST';
  return this.run(options, callback);
};

DataFluxFunc.prototype.put = function(options, callback) {
  options = options || {};
  options.method = 'PUT';
  return this.run(options, callback);
};

DataFluxFunc.prototype.delete = function(options, callback) {
  options = options || {};
  options.method = 'DELETE';
  return this.run(options, callback);
};

exports.DataFluxFunc = DataFluxFunc;

if (require.main === module) {
  var host = process.argv[2] || 'localhost:8088';

  // Create DataFlux Func Handler
  var opt = {
    akId    : 'ak-xxxxx',
    akSecret: 'xxxxxxxxxx',
    host    : host,
  };
  var dff = new DataFluxFunc(opt);

  // Debug ON
  dff.debug = true;

  // Send GET Request
  var getOpt = {
    path: '/api/v1/do/ping',
  };
  dff.get(getOpt, function(err, respData, respStatusCode) {
    if (err) console.error(colored(err, 'red'))

    // Send POST Request
    var postOpt = {
      path: '/api/v1/do/echo',
      body: {
        'echo': {
          'int'    : 1,
          'str'    : 'Hello World',
          'unicode': '你好，世界！',
          'none'   : null,
          'boolean': true,
        }
      }
    };
    dff.post(postOpt, function(err, respData, respStatusCode) {
      if (err) console.error(colored(err, 'red'))

      // Send UPLOAD Request
      if (UPLOAD_DISABLED) return;

      var filename = __filename.split('/').pop();
      var uploadOpt = {
        path      : '/api/v1/resources/do/upload',
        fileBuffer: fs.readFileSync(filename),
        filename  : filename,
        fields    : {
          'folder': 'test'
        },
      };
      dff.upload(uploadOpt, function(err, respData, respStatusCode) {
        if (err) console.error(colored(err, 'red'))
      });
    });
  });
};
